File: gs/gs4719/week03/main.py
# 匯入 keyword 模組
import keyword

# 列印關鍵字列表
print(keyword.kwlist)

# 檢查 'if' 是否是關鍵字，預期結果是 True
is_if_keyword = keyword.iskeyword('if')
print("Is 'if' a keyword?", is_if_keyword)

# 檢查 'IF' 是否是關鍵字，預期結果是 False
is_IF_keyword = keyword.iskeyword('IF')
print("Is 'IF' a keyword?", is_IF_keyword)

"""
"""

# 整數相加
result1 = 23 + 45
print(result1)

# 整數相減
result2 = 59 - 22
print(result2)

# 整數相乘
result3 = 43 * 3
print(result3)

# 整數平方
result4 = 8 ** 2
print(result4)

# 整數除法（浮點數結果）
result5 = 16 / 4
print(result5)

# 整數除法（整數結果）
result6 = 16 // 4
print(result6)

# 整數除法（浮點數結果）
result7 = 16 // 4.0
print(result7)

# 取餘數
result8 = 15 % 4
print(result8)

# 取餘數（浮點數結果）
result9 = 15 % 4.0
print(result9)

# 定義變數 x 並賦值
x = 5
print(x)

# 增加 x 的值
x = x + 1
print(x)

# 增加 x 的值（簡化語法）
x += 1
print(x)

# 遞減 x 的值
x -= 1
print(x)

# 負負得正
x = -x
print(x)

# 不支援後置遞減運算符 x--
# 定義變數 y 並賦值
y = 3
print(x, y)

# 將 y 的值提高到 x 的次方
y **= x
print(x, y)

# 大數計算
big_number1 = 33 ** 999
print(big_number1)

# 大數計算（浮點數結果）
big_number2 = 33.0 ** 999
print(big_number2)

"""
"""

# 整數 36
print(36)

# 二進位數字 0b1011，表示十進位的 11
print(0b1011)

# 八進位數字 0o36，表示十進位的 30
print(0o36)

# 十六進位數字 0x36，表示十進位的 54
print(0x36)

# 加法操作，包括十進位、二進位、八進位和十六進位
result = 11 + 0b11 + 0o11 + 0x11
print(result)

# 浮點數 31.2
print(312e-2)

# 將浮點數轉換為整數，結果是 31
int_result = int(312e-2)
print(int_result)

# 整數 20
print(20)

# 將整數轉換為浮點數，結果是 20.0
float_result = float(20)
print(float_result)

# 將浮點數 25.346 保留兩位小數，結果是 25.35
rounded_result = round(25.346, 2)
print(rounded_result)

# 定義整數變數 x，值為 33
x = 33
print("type(x) =", type(x))
print("id(x) =", id(x))
print("x =", x)

# 將變數 x 的值更改為浮點數 63.6
x = 63.6
print("type(x) =", type(x))
print("id(x) =", id(x))
print("x =", x)

print("abc".isidentifier()) # True
print("99a".isidentifier()) # False
print("_".isidentifier()) # True
print("for".isidentifier()) # True

File: gs/gs4719/week03/main.go
package main

import (
	"fmt"
	"math"
)

func main() {
    // Checking if 'if' is a keyword
    isIfKeyword := isKeyword("if")
    fmt.Printf("Is 'if' a keyword? %t\n", isIfKeyword)

    // Checking if 'IF' is a keyword
    isIFKeyword := isKeyword("IF")
    fmt.Printf("Is 'IF' a keyword? %t\n", isIFKeyword)

    // Integer addition
    result1 := 23 + 45
    fmt.Println(result1)

    // Integer subtraction
    result2 := 59 - 22
    fmt.Println(result2)

    // Integer multiplication
    result3 := 43 * 3
    fmt.Println(result3)

    // Integer square
    result4 := 8 * 8 // Integer square
    fmt.Println(result4)

    // Integer division (floating-point result)
    result5 := float64(16) / 4
    fmt.Println(result5)

    // Integer division (integer result)
    result6 := 16 / 4
    fmt.Println(result6)

    // Integer division (floating-point result)
    result7 := float64(16) / 4.0
    fmt.Println(result7)

    // Modulus
    result8 := 15 % 4
    fmt.Println(result8)

    // Modulus (floating-point result)
    result9 := float64(15) / 4.0
    fmt.Println(result9)

    // Defining and initializing a variable x
    x := 5
    fmt.Println(x)

    // Increasing the value of x
    x = x + 1
    fmt.Println(x)

    // Increasing the value of x (simplified syntax)
    x++
    fmt.Println(x)

    // Decreasing the value of x
    x--
    fmt.Println(x)

    // Negating x
    x = -x
    fmt.Println(x)

    // Defining and initializing a variable y
    y := 3
    fmt.Println(x, y)

    // Raising y to the power of x
    y = int(math.Pow(float64(x), float64(y)))
    fmt.Println(x, y)

    // Large number calculation
    bigNumber1 := math.Pow(33, 999)
    fmt.Println(bigNumber1)

    // Large number calculation (floating-point result)
    bigNumber2 := math.Pow(33.0, 999)
    fmt.Println(bigNumber2)

    // Integers
    fmt.Println(36)

    // Binary number 0b1011, equivalent to decimal 11
    fmt.Println(11)

    // Octal number 0o36, equivalent to decimal 30
    fmt.Println(30)

    // Hexadecimal number 0x36, equivalent to decimal 54
    fmt.Println(54)

    // Addition with different number representations
    result10 := 11 + 3 + 9 + 17
    fmt.Println(result10)

    // Floating-point number 31.2
    fmt.Println(31.2)

    // Converting a floating-point number to an integer (result is 31)
	intResult := int(math.Round(31.2))
    fmt.Println(intResult)

    // Integer 20
    fmt.Println(20)

    // Converting an integer to a floating-point number (result is 20.0)
    floatResult := float64(20)
    fmt.Println(floatResult)

    // Rounding a floating-point number to two decimal places (result is 25.35)
    roundedResult := round(25.346, 2)
    fmt.Println(roundedResult)

    // Defining an integer variable x with a value of 33
    x = 33
    fmt.Printf("type(x) = %T\n", x)
    fmt.Printf("x = %v\n", x)

    // Identifiers in Go
    fmt.Println(isIdentifier("abc"))  // true
    fmt.Println(isIdentifier("99a"))  // false
    fmt.Println(isIdentifier("_"))    // true
    fmt.Println(isIdentifier("for"))  // true
}

// isKeyword checks if a string is a keyword in Go
func isKeyword(str string) bool {
	// List of Go keywords
	keywords := map[string]bool{
		"break":       true,
		"default":     true,
		"func":        true,
		"interface":   true,
		"select":      true,
		"case":        true,
		"defer":       true,
		"go":          true,
		"map":         true,
		"struct":      true,
		"chan":        true,
		"else":        true,
		"goto":        true,
		"package":     true,
		"switch":      true,
		"const":       true,
		"fallthrough": true,
		"if":          true,
		"range":       true,
		"type":        true,
		"continue":    true,
		"for":         true,
		"import":      true,
		"return":      true,
		"var":         true,
	}

	return keywords[str]
}

// isIdentifier checks if a string is a valid identifier in Go
func isIdentifier(str string) bool {
	if len(str) == 0 {
		return false
	}

	for _, char := range str {
		if !isLetter(char) && !isDigit(char) && char != '_' {
			return false
		}
	}

	// 這裡添加了對第一個字符是否為字母的檢查
	if !isLetter(rune(str[0])) {
		return false
	}

	return true
}


// isLetter checks if a character is a letter (uppercase or lowercase)
func isLetter(char rune) bool {
	return ('a' <= char && char <= 'z') || ('A' <= char && char <= 'Z')
}

// isDigit checks if a character is a digit
func isDigit(char rune) bool {
	return '0' <= char && char <= '9'
}

// round rounds a floating-point number to a specified number of decimal places
func round(num float64, decimalPlaces int) float64 {
	shift := math.Pow(10, float64(decimalPlaces))
	return math.Round(num*shift) / shift
}
